### PR TITLE
kvlds-dump/undump utilities

### PR DESCRIPTION
**Pull request by cperciva**

### CODE REVIEW

File: kvlds-undump/main.c [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+		goto err2;
+	}
+	if (sb.st_size > 255) {
+		warn0("File is too large (%ju bytes): %s",
+		    (intmax_t)(sb.st_size), s);
+		goto err2;
+	}
+
+	/* Read value. */
+	if (fread(buf, sb.st_size, 1, f) != 1) {
```
Comment (gperciva) [inline kvlds-undump/main.c]:
```
if (fread(buf, (size_t)sb.st_size, 1, f) != 1) {
```

File: kvlds-undump/main.c [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+		warnp("fread(%s)", s);
+		goto err2;
+	}
+
+	/* Construct kvlds key. */
+	if ((k = kvldskey_create(buf, sb.st_size)) == NULL)
```
Comment (gperciva) [inline kvlds-undump/main.c]:
```
if ((k = kvldskey_create(buf, (size_t)sb.st_size)) == NULL)
```

Comment (cperciva) [reply kvlds-undump/main.c]: Thanks, fixed.

**Comment by cperciva** [comment]
Any thoughts on the test code?

**Comment by gperciva** [comment]
No thoughts yet, might be a few days.

File: kvlds-dump/main.c [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+			usage();
+		}
+	}
+	argc -= optind;
+	argv += optind;
+
+	/* We should have processed all the arguments. */
+	if (argc != 0)
+		usage();
+
```
Comment (gperciva) [inline kvlds-dump/main.c]: clang static analyzer would be happier with:
```
(void)argv; /* argv is not used beyond this point. */
```

**Comment by cperciva** [comment]
No problem, just wanted to make sure the pause after your first two comments didn't mean that you were finished.

File: lib/util/kivaloo.c [hunk 4]
```diff
@@ -0,0 +1,89 @@
+#include <stdlib.h>
+#include <unistd.h>
+
+#include "sock.h"
+#include "warnp.h"
+#include "wire.h"
+
+#include "kivaloo.h"
+
+struct kivaloo_cookie {
```
Comment (gperciva) [inline lib/util/kivaloo.c]: At first glance, I don't see anything kivaloo-specific in here.  Couldn't it be used for `lbs` as well?  I'd be tempted to:
- rename it to something like `wire_wrapper` (if not include it directly in `lib/wire/wire.h`).
- use it in programs like `tests/lbs/` and `tests/kvlds`.

**Comment by gperciva** [comment]
This is relevant to your interests: https://github.com/Tarsnap/libcperciva/pull/170

File: lib/util/kivaloo.c [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	return (NULL);
+}
+
+/**
+ * kivaloo_close(cookie):
+ * Destroy and free the wire request queue, close the socket and free memory
+ * allocated by the kivaloo_open which returned ${cookie}.
+ */
+void
+kivaloo_close(void * cookie)
```
Comment (gperciva) [inline lib/util/kivaloo.c]: Do you want this to be `kivaloo_close()` or `kivaloo_free()`?  Namely, should it behave if given `NULL`?

Comment (cperciva) [reply lib/util/kivaloo.c]: I think open/close makes more sense.  So no, it doesn't need to accept NULL.

**Comment by cperciva** [comment]
@gperciva Can you walk me through fixing this branch now that some of its commits are already on master?

**Comment by gperciva** [comment]
I don't see any conflicts, so this is trivial.

```
git checkout master
git pull -r
git checkout utils
git rebase master
```

... huh, apparently the first round of "add missing includes" didn't cover this case.

**Comment by cperciva** [comment]
Ok, I followed those instructions and now I get:

```
[cperciva@exbuntu /usr/home/cperciva/tarsnap/tarsnap-git/kivaloo]$ git push origin utils
To github.com:Tarsnap/kivaloo.git
 ! [rejected]        utils -> utils (non-fast-forward)
error: failed to push some refs to 'git@github.com:Tarsnap/kivaloo.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.
```

**Comment by gperciva** [comment]
That's because github has a different history for its version.  Add a `--force`, i.e.
```
git push --force origin utils
```

(on my system, I omit the last two words, but it depends on your version of git and what the defautl push behaviour is.

**Comment by cperciva** [comment]
Ah, ok.  I was thinking that "rebase on top of the latest tree" was a common enough operation that it wouldn't need `--force`.

File: kvlds-dump/DESIGN [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+=================
+
+The kvlds-dump utility is invoked as
+# kivaloo-kvlds-dump -t <kvlds socket> [--fs <dir>]
+
+It opens a connection to the specified KVLDS daemon and retrieves all of the
+key-value pairs.  By default the key-value pairs are written to stdout; if
+the --fs <dir> option is specified, numbered directories will be created
+under that directory containing files named "k" and "v" with the keys and
+values respectively.
```
Comment (gperciva) [inline kvlds-dump/DESIGN]: Is it worth adding something like "the directory must exist" ?  Or do you think it's worth adding a `mkdir()` in kvlds-dump?

I'm not certain whether it's more customary to create a dir or not, but I was surprised when I tried to run it and was told
```
kvlds-dump: chdir(dir): No such file or directory
```

File: kvlds-dump/main.c [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+		    "Dumped %" PRIu64 " key-value pairs in %f seconds.",
+		    C.N, timeval_diff(en, st));
+	}
+
+	/* Free memory allocated by kvldskey_create. */
+	kvldskey_free(nullkey);
+
+	/* Close the connection to KVLDS. */
+	kivaloo_close(K);
+
```
Comment (gperciva) [inline kvlds-dump/main.c]: When I move this into a separate PR, I'll add an

```
events_shutdown()
```
here to make valgrind happy.

**Comment by gperciva** [comment]
Checking

```
git checkout master
git pull -r
git diff HEAD..utils
```

shows that all the changes in this branch are now obsolete.

**Pull request closed without merging**